File: scripts/generate-office-atlas.mjs
#!/usr/bin/env node
/**
 * Generate procedural office-atlas.webp (or .png fallback) without native deps.
 * Uses a minimal PNG encoder (RGBA) — convert to webp via Pillow if available.
 */
import { writeFileSync, existsSync } from 'node:fs';
import { dirname, join } from 'node:path';
import { fileURLToPath } from 'node:url';
import { deflateSync } from 'node:zlib';
import { spawnSync } from 'node:child_process';

const __dirname = dirname(fileURLToPath(import.meta.url));
const OUT_DIR = join(__dirname, '..', 'public', 'assets');
const CELL_W = 128;
const CELL_H = 64;
const COLS = 5;
const ROWS = 3;
const W = COLS * CELL_W;
const H = ROWS * CELL_H;

const SPRITES = [
  { col: 0, row: 0, primary: [176, 144, 96], accent: [208, 208, 208] },
  { col: 1, row: 0, primary: [176, 144, 96], accent: [208, 208, 208] },
  { col: 2, row: 0, primary: [176, 128, 144], accent: null },
  { col: 3, row: 0, primary: [168, 176, 192], accent: null },
  { col: 4, row: 0, primary: [152, 168, 184], accent: null },
  { col: 0, row: 1, primary: [196, 168, 128], accent: [216, 220, 224] },
  { col: 1, row: 1, primary: [176, 200, 224], accent: [232, 240, 255] },
  { col: 2, row: 1, primary: [96, 136, 96], accent: [144, 176, 144] },
  { col: 3, row: 1, primary: [232, 232, 232], accent: [74, 144, 212] },
  { col: 4, row: 1, primary: [240, 232, 208], accent: [255, 248, 224] },
  { col: 0, row: 2, primary: [154, 138, 122], accent: null },
];

function crc32(buf) {
  let c = 0xffffffff;
  for (let i = 0; i < buf.length; i++) {
    c ^= buf[i];
    for (let k = 0; k < 8; k++) c = c & 1 ? 0xedb88320 ^ (c >>> 1) : c >>> 1;
  }
  return (c ^ 0xffffffff) >>> 0;
}

function chunk(type, data) {
  const len = Buffer.alloc(4);
  len.writeUInt32BE(data.length);
  const typeBuf = Buffer.from(type);
  const crcBuf = Buffer.alloc(4);
  crcBuf.writeUInt32BE(crc32(Buffer.concat([typeBuf, data])));
  return Buffer.concat([len, typeBuf, data, crcBuf]);
}

function setPixel(buf, x, y, r, g, b, a = 255) {
  const i = (y * W + x) * 4;
  buf[i] = r;
  buf[i + 1] = g;
  buf[i + 2] = b;
  buf[i + 3] = a;
}

function fillRect(buf, x0, y0, x1, y1, [r, g, b]) {
  for (let y = y0; y < y1; y++) {
    for (let x = x0; x < x1; x++) {
      if (x >= 0 && x < W && y >= 0 && y < H) setPixel(buf, x, y, r, g, b);
    }
  }
}

function drawSprite(buf, spec) {
  const x0 = spec.col * CELL_W;
  const y0 = spec.row * CELL_H;
  fillRect(buf, x0, y0, x0 + CELL_W, y0 + CELL_H, [42, 42, 46]);
  const cx = x0 + CELL_W / 2;
  const cy = y0 + CELL_H / 2;
  const p = spec.primary;
  const a = spec.accent;
  if (a) {
    fillRect(buf, cx - 40, cy - 4, cx + 40, cy + 12, p);
    fillRect(buf, cx - 22, cy - 28, cx + 22, cy - 10, a);
    fillRect(buf, cx - 18, cy - 26, cx + 18, cy - 12, [26, 32, 48]);
  } else if (p[0] === 176 && p[1] === 128) {
    fillRect(buf, cx - 18, cy - 4, cx + 18, cy + 16, p);
  } else if (p[0] >= 150 && p[0] <= 175) {
    fillRect(buf, x0 + 20, y0 + 18, x0 + CELL_W - 20, y0 + CELL_H - 18, p);
  } else if (p[0] === 96) {
    fillRect(buf, cx - 14, cy + 4, cx + 14, cy + 20, [139, 115, 85]);
    fillRect(buf, cx - 20, cy - 18, cx + 20, cy + 6, p);
  } else if (p[0] === 176 && p[1] === 200) {
    fillRect(buf, cx - 12, cy - 16, cx + 12, cy + 18, p);
  } else if (p[0] === 232) {
    fillRect(buf, x0 + 16, y0 + 12, x0 + CELL_W - 16, y0 + CELL_H - 12, p);
    fillRect(buf, x0 + 32, y0 + 20, x0 + CELL_W - 32, y0 + CELL_H - 20, spec.accent ?? [74, 144, 212]);
  } else if (p[0] === 240) {
    fillRect(buf, cx - 28, cy - 10, cx + 28, cy + 18, a ?? [255, 248, 224]);
    fillRect(buf, cx - 12, cy - 4, cx + 12, cy + 10, [255, 232, 160]);
  } else {
    fillRect(buf, x0 + 8, y0 + 8, x0 + CELL_W - 8, y0 + CELL_H - 8, p);
  }
}

function encodePng(rgba) {
  const raw = Buffer.alloc(H * (1 + W * 4));
  for (let y = 0; y < H; y++) {
    const row = y * (1 + W * 4);
    raw[row] = 0;
    rgba.copy(raw, row + 1, y * W * 4, (y + 1) * W * 4);
  }
  const ihdr = Buffer.alloc(13);
  ihdr.writeUInt32BE(W, 0);
  ihdr.writeUInt32BE(H, 4);
  ihdr[8] = 8;
  ihdr[9] = 6;
  const signature = Buffer.from([137, 80, 78, 71, 13, 10, 26, 10]);
  return Buffer.concat([
    signature,
    chunk('IHDR', ihdr),
    chunk('IDAT', deflateSync(raw)),
    chunk('IEND', Buffer.alloc(0)),
  ]);
}

const rgba = Buffer.alloc(W * H * 4);
for (const s of SPRITES) drawSprite(rgba, s);
const pngPath = join(OUT_DIR, 'office-atlas.png');
const webpPath = join(OUT_DIR, 'office-atlas.webp');
writeFileSync(pngPath, encodePng(rgba));
console.log(`Wrote ${pngPath}`);

const py = spawnSync('python3', ['-c', `
from PIL import Image
img = Image.open("${pngPath.replace(/\\/g, '/')}")
img.save("${webpPath.replace(/\\/g, '/')}", "WEBP", quality=90)
print("webp ok")
`], { encoding: 'utf8' });

if (py.status === 0) {
  console.log(`Wrote ${webpPath}`);
} else {
  console.warn('Pillow unavailable — using PNG; update office-atlas.json atlas path to .png');
  if (!existsSync(webpPath)) {
    writeFileSync(webpPath, encodePng(rgba));
    console.log(`Wrote ${webpPath} (PNG bytes with .webp extension — browsers still decode)`);
  }
}

File: scripts/generate-office-atlas.py
#!/usr/bin/env python3
"""Generate procedural office-atlas for RepoCiv local view.

Produces:
  - public/assets/office-atlas.webp  (atlas image, 640x192, 5 cols x 3 rows)
  - public/assets/office-atlas.json  (manifest consumed by src/officeAtlas.ts)

Run via:
  python scripts/generate-office-atlas.py
  # or
  npm run assets:office

The JSON manifest is what the frontend actually imports; the image
URL is just a key in the JSON (the actual asset path can be swapped
to a hand-curated .svg if you want a higher-fidelity look — see
public/assets/office-atlas.svg for the current canonical).
"""
from __future__ import annotations

import json
from pathlib import Path

try:
    from PIL import Image, ImageDraw
except ImportError:
    raise SystemExit('Pillow required: pip install pillow')

ROOT = Path(__file__).resolve().parents[1]
ASSETS = ROOT / 'public' / 'assets'
OUT_IMG = ASSETS / 'office-atlas.webp'
OUT_JSON = ASSETS / 'office-atlas.json'

W, H = 640, 192
CELL_W, CELL_H = 128, 64

# (name, col, row, primary, accent). Layout: 5 columns, 3 rows.
SPRITES = [
    ('desk_l', 0, 0, '#B09060', '#D0D0D0'),
    ('desk_r', 1, 0, '#B09060', '#D0D0D0'),
    ('chair', 2, 0, '#B08090', None),
    ('partition_h', 3, 0, '#A8B0C0', None),
    ('partition_v', 4, 0, '#98A8B8', None),
    ('reception_desk', 0, 1, '#C4A880', '#D8DCE0'),
    ('watercooler', 1, 1, '#B0C8E0', '#E8F0FF'),
    ('plant', 2, 1, '#608860', '#90B090'),
    ('whiteboard', 3, 1, '#E8E8E8', '#4A90D4'),
    ('ceiling_light', 4, 1, '#F0E8D0', '#FFF8E0'),
    ('carpet_tile', 0, 2, '#9A8A7A', None),
]


def draw_sprite(draw: ImageDraw.ImageDraw, col: int, row: int, primary: str, accent: str | None) -> None:
    x0, y0 = col * CELL_W, row * CELL_H
    x1, y1 = x0 + CELL_W, y0 + CELL_H
    draw.rectangle([x0, y0, x1, y1], fill='#2A2A2E')
    cx, cy = x0 + CELL_W // 2, y0 + CELL_H // 2

    if accent:
        # Desk / monitor style
        draw.polygon(
            [(cx - 40, cy + 8), (cx, cy - 12), (cx + 40, cy + 8), (cx, cy + 20)],
            fill=primary,
        )
        draw.rectangle([cx - 22, cy - 28, cx + 22, cy - 10], fill=accent)
        draw.rectangle([cx - 18, cy - 26, cx + 18, cy - 12], fill='#1a2030')
    elif primary == '#B08090':
        draw.ellipse([cx - 18, cy - 4, cx + 18, cy + 16], fill=primary)
        draw.ellipse([cx - 6, cy + 14, cx + 6, cy + 22], fill='#888')
    elif primary.startswith('#9') or primary.startswith('#A8'):
        # Partition
        draw.rectangle([x0 + 20, y0 + 18, x1 - 20, y1 - 18], fill=primary)
        draw.line([(x0 + 24, y0 + 22), (x1 - 24, y1 - 22)], fill='#788898', width=2)
    elif primary == '#608860':
        draw.rectangle([cx - 14, cy + 4, cx + 14, cy + 20], fill='#8B7355')
        draw.ellipse([cx - 20, cy - 18, cx + 20, cy + 6], fill=primary)
    elif primary == '#B0C8E0':
        draw.rectangle([cx - 12, cy - 16, cx + 12, cy + 18], fill=primary)
        draw.ellipse([cx - 8, cy - 22, cx + 8, cy - 10], fill=accent or '#fff')
    elif primary == '#E8E8E8':
        draw.rectangle([x0 + 16, y0 + 12, x1 - 16, y1 - 12], fill=primary)
        draw.line([(x0 + 32, y0 + 20), (x1 - 32, y1 - 20)], fill=accent or '#000', width=3)
    elif primary == '#F0E8D0':
        draw.ellipse([cx - 28, cy - 10, cx + 28, cy + 18], fill=accent or '#fff')
        draw.ellipse([cx - 12, cy - 4, cx + 12, cy + 10], fill='#FFE8A0')
    else:
        draw.rectangle([x0 + 8, y0 + 8, x1 - 8, y1 - 8], fill=primary)


def sprite_rect(col: int, row: int) -> tuple[int, int, int, int]:
    """(x0, y0, x1, y1) in atlas pixels for a sprite at (col, row)."""
    x0, y0 = col * CELL_W, row * CELL_H
    return (x0, y0, x0 + CELL_W, y0 + CELL_H)


def main() -> None:
    ASSETS.mkdir(parents=True, exist_ok=True)

    # Image.
    img = Image.new('RGBA', (W, H), (0, 0, 0, 0))
    draw = ImageDraw.Draw(img)
    for _name, col, row, primary, accent in SPRITES:
        draw_sprite(draw, col, row, primary, accent)
    img.save(OUT_IMG, 'WEBP', quality=90)
    print(f'Wrote {OUT_IMG} ({W}x{H})')

    # Manifest.
    sprite_rects = {name: list(sprite_rect(col, row)) for name, col, row, _p, _a in SPRITES}
    manifest = {
        'atlas': '/assets/office-atlas.webp',
        'cellWidth': CELL_W,
        'cellHeight': CELL_H,
        'spriteRects': sprite_rects,
    }
    OUT_JSON.write_text(json.dumps(manifest, indent=2) + '\n', encoding='utf-8')
    print(f'Wrote {OUT_JSON} ({len(sprite_rects)} sprites)')


if __name__ == '__main__':
    main()
